File: js/src/core/e2ee.ts
import { gcm } from '@noble/ciphers/aes';
import { xchacha20poly1305 } from '@noble/ciphers/chacha';
import { hkdf } from '@noble/hashes/hkdf';
import { sha256 } from '@noble/hashes/sha256';
import { Buffer } from 'buffer';
import ed2curve from 'ed2curve';
import { ethers } from 'ethers';
import nacl from 'tweetnacl';
import type { SigningAlgo } from '../types/attestation-common';
import { hexToBuffer, utf8 } from '../utils/common';
import { VerificationError } from '../utils/errors';

const E2EE_KEY_BYTES = 32;
const ED25519_NONCE_BYTES = 24;
const ED25519_ENVELOPE_PREFIX_BYTES = E2EE_KEY_BYTES + ED25519_NONCE_BYTES;
const ECDSA_PUBLIC_KEY_BYTES = 65;
const ECDSA_RAW_PUBLIC_KEY_BYTES = 64;
const ECDSA_NONCE_BYTES = 12;
const ECDSA_ENVELOPE_PREFIX_BYTES = ECDSA_PUBLIC_KEY_BYTES + ECDSA_NONCE_BYTES;
const E2EE_AUTH_TAG_BYTES = 16;
const ED25519_HKDF_INFO = utf8('ed25519_encryption');
const ECDSA_HKDF_INFO = utf8('ecdsa_encryption');

/** A quote-bound model key selected for the E2EE protocol. */
export type E2eeModelKey = {
  readonly signingAlgo: SigningAlgo;
  readonly publicKey: string;
};

type Ed25519E2eeClientKeyPair = {
  readonly signingAlgo: 'ed25519';
  readonly publicKey: string;
  readonly x25519SecretKey: Uint8Array;
};

type EcdsaE2eeClientKeyPair = {
  readonly signingAlgo: 'ecdsa';
  /** A 64-byte uncompressed secp256k1 X || Y public key. */
  readonly publicKey: string;
  readonly privateKey: string;
};

export type E2eeClientKeyPair =
  | Ed25519E2eeClientKeyPair
  | EcdsaE2eeClientKeyPair;

type EncryptE2eeTextParams = {
  readonly plaintext: string;
  readonly modelKey: E2eeModelKey;
};

type DecryptE2eeTextParams = {
  readonly ciphertext: string;
  readonly clientKeyPair: E2eeClientKeyPair;
  readonly field: string;
};

/** Create a fresh client key pair for the selected E2EE protocol. */
export function createE2eeClientKeyPair(
  signingAlgo: SigningAlgo = 'ed25519',
): E2eeClientKeyPair {
  if (signingAlgo === 'ecdsa') {
    const privateKey = `0x${Buffer.from(nacl.randomBytes(E2EE_KEY_BYTES)).toString('hex')}`;
    const signingKey = new ethers.SigningKey(privateKey);
    return {
      signingAlgo,
      publicKey: signingKey.publicKey.slice(4),
      privateKey,
    };
  }

  const keyPair = nacl.sign.keyPair();
  const x25519SecretKey = ed2curve.convertSecretKey(keyPair.secretKey);
  if (x25519SecretKey === null) {
    throw new VerificationError({ code: 'e2ee.model_public_key_invalid' });
  }
  return {
    signingAlgo,
    publicKey: Buffer.from(keyPair.publicKey).toString('hex'),
    x25519SecretKey,
  };
}

/** Encrypt one UTF-8 field for a quote-bound model E2EE key. */
export function encryptE2eeText({
  plaintext,
  modelKey,
}: EncryptE2eeTextParams): string {
  return modelKey.signingAlgo === 'ed25519'
    ? encryptEd25519Text({ plaintext, publicKey: modelKey.publicKey })
    : encryptEcdsaText({ plaintext, publicKey: modelKey.publicKey });
}

/** Decrypt one UTF-8 field produced by the selected E2EE protocol. */
export function decryptE2eeText({
  ciphertext,
  clientKeyPair,
  field,
}: DecryptE2eeTextParams): string {
  // The server represents an encrypted empty field as an empty string rather
  // than an envelope. Preserve that protocol value without treating it as a
  // malformed ciphertext.
  if (ciphertext === '') {
    return '';
  }
  try {
    return clientKeyPair.signingAlgo === 'ed25519'
      ? decryptEd25519Text({ ciphertext, clientKeyPair, field })
      : decryptEcdsaText({ ciphertext, clientKeyPair, field });
  } catch (cause) {
    throw new VerificationError(
      {
        code: 'e2ee.decryption_failed',
        details: { field },
      },
      { cause },
    );
  }
}

type EncryptProtocolTextParams = {
  readonly plaintext: string;
  readonly publicKey: string;
};

function encryptEd25519Text({
  plaintext,
  publicKey,
}: EncryptProtocolTextParams): string {
  const recipientX25519PublicKey = toX25519PublicKey(publicKey);
  const ephemeralSecretKey = nacl.randomBytes(E2EE_KEY_BYTES);
  const ephemeralPublicKey = nacl.scalarMult.base(ephemeralSecretKey);
  const nonce = nacl.randomBytes(ED25519_NONCE_BYTES);
  const encryptionKey = deriveE2eeKey({
    sharedSecret: nacl.scalarMult(ephemeralSecretKey, recipientX25519PublicKey),
    info: ED25519_HKDF_INFO,
  });
  const ciphertext = xchacha20poly1305(encryptionKey, nonce).encrypt(
    utf8(plaintext),
  );

  return Buffer.concat([
    Buffer.from(ephemeralPublicKey),
    Buffer.from(nonce),
    Buffer.from(ciphertext),
  ]).toString('hex');
}

function encryptEcdsaText({
  plaintext,
  publicKey,
}: EncryptProtocolTextParams): string {
  const recipientPublicKey = toEcdsaPublicKey(publicKey);
  const ephemeralPrivateKey = `0x${Buffer.from(nacl.randomBytes(E2EE_KEY_BYTES)).toString('hex')}`;
  const ephemeralKey = new ethers.SigningKey(ephemeralPrivateKey);
  const nonce = nacl.randomBytes(ECDSA_NONCE_BYTES);
  const encryptionKey = deriveE2eeKey({
    sharedSecret: ecdsaSharedSecret({
      privateKey: ephemeralPrivateKey,
      publicKey: recipientPublicKey,
    }),
    info: ECDSA_HKDF_INFO,
  });
  const ciphertext = gcm(encryptionKey, nonce).encrypt(utf8(plaintext));

  return Buffer.concat([
    Buffer.from(ephemeralKey.publicKey.slice(2), 'hex'),
    Buffer.from(nonce),
    Buffer.from(ciphertext),
  ]).toString('hex');
}

type DecryptProtocolTextParams<TClientKeyPair extends E2eeClientKeyPair> = {
  readonly ciphertext: string;
  readonly clientKeyPair: TClientKeyPair;
  readonly field: string;
};

function decryptEd25519Text({
  ciphertext,
  clientKeyPair,
  field,
}: DecryptProtocolTextParams<Ed25519E2eeClientKeyPair>): string {
  const envelope = hexToBuffer(ciphertext, field);
  if (envelope.length < ED25519_ENVELOPE_PREFIX_BYTES + E2EE_AUTH_TAG_BYTES) {
    throw new Error('E2EE envelope is too short');
  }
  const ephemeralPublicKey = envelope.subarray(0, E2EE_KEY_BYTES);
  const nonce = envelope.subarray(
    E2EE_KEY_BYTES,
    ED25519_ENVELOPE_PREFIX_BYTES,
  );
  const encrypted = envelope.subarray(ED25519_ENVELOPE_PREFIX_BYTES);
  const decryptionKey = deriveE2eeKey({
    sharedSecret: nacl.scalarMult(
      clientKeyPair.x25519SecretKey,
      ephemeralPublicKey,
    ),
    info: ED25519_HKDF_INFO,
  });
  const plaintext = xchacha20poly1305(decryptionKey, nonce).decrypt(encrypted);
  return new TextDecoder('utf-8', { fatal: true }).decode(plaintext);
}

function decryptEcdsaText({
  ciphertext,
  clientKeyPair,
  field,
}: DecryptProtocolTextParams<EcdsaE2eeClientKeyPair>): string {
  const envelope = hexToBuffer(ciphertext, field);
  if (envelope.length < ECDSA_ENVELOPE_PREFIX_BYTES + E2EE_AUTH_TAG_BYTES) {
    throw new Error('E2EE envelope is too short');
  }
  const ephemeralPublicKey = envelope.subarray(0, ECDSA_PUBLIC_KEY_BYTES);
  if (ephemeralPublicKey[0] !== 0x04) {
    throw new Error('ECDSA ephemeral public key is not uncompressed');
  }
  const nonce = envelope.subarray(
    ECDSA_PUBLIC_KEY_BYTES,
    ECDSA_ENVELOPE_PREFIX_BYTES,
  );
  const encrypted = envelope.subarray(ECDSA_ENVELOPE_PREFIX_BYTES);
  const decryptionKey = deriveE2eeKey({
    sharedSecret: ecdsaSharedSecret({
      privateKey: clientKeyPair.privateKey,
      publicKey: `0x${ephemeralPublicKey.toString('hex')}`,
    }),
    info: ECDSA_HKDF_INFO,
  });
  const plaintext = gcm(decryptionKey, nonce).decrypt(encrypted);
  return new TextDecoder('utf-8', { fatal: true }).decode(plaintext);
}

function toX25519PublicKey(publicKey: string): Uint8Array {
  const ed25519PublicKey = hexToBuffer(publicKey, 'modelSigningPublicKey');
  if (ed25519PublicKey.length !== E2EE_KEY_BYTES) {
    throw new VerificationError({ code: 'e2ee.model_public_key_invalid' });
  }
  const x25519PublicKey = ed2curve.convertPublicKey(ed25519PublicKey);
  if (x25519PublicKey === null) {
    throw new VerificationError({ code: 'e2ee.model_public_key_invalid' });
  }
  return x25519PublicKey;
}

function toEcdsaPublicKey(publicKey: string): string {
  const bytes = hexToBuffer(publicKey, 'modelSigningPublicKey');
  if (bytes.length === ECDSA_RAW_PUBLIC_KEY_BYTES) {
    return `0x04${bytes.toString('hex')}`;
  }
  if (bytes.length === ECDSA_PUBLIC_KEY_BYTES && bytes[0] === 0x04) {
    return `0x${bytes.toString('hex')}`;
  }
  throw new VerificationError({ code: 'e2ee.model_public_key_invalid' });
}

type DeriveE2eeKeyParams = {
  readonly sharedSecret: Uint8Array;
  readonly info: Uint8Array;
};

function deriveE2eeKey({
  sharedSecret,
  info,
}: DeriveE2eeKeyParams): Uint8Array {
  return hkdf(sha256, sharedSecret, undefined, info, E2EE_KEY_BYTES);
}

type EcdsaSharedSecretParams = {
  readonly privateKey: string;
  readonly publicKey: string;
};

/** The legacy ECDSA protocol uses the X coordinate of the shared point. */
function ecdsaSharedSecret({
  privateKey,
  publicKey,
}: EcdsaSharedSecretParams): Uint8Array {
  const sharedPoint = ethers.getBytes(
    new ethers.SigningKey(privateKey).computeSharedSecret(publicKey),
  );
  return sharedPoint.subarray(1, E2EE_KEY_BYTES + 1);
}

File: py/src/nearai_inference_sdk/core/e2ee.py
"""Field encryption for the Gateway's Ed25519-v2 and legacy ECDSA protocols."""

from __future__ import annotations

import secrets
from dataclasses import dataclass
from typing import Literal

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from nacl import bindings
from nacl.signing import SigningKey

from ..types.attestation_common import SigningAlgo
from ..types.e2ee import E2eeModelKey
from ..utils.common import hex_to_bytes
from ..utils.errors import verification_failure


@dataclass(frozen=True, kw_only=True)
class Ed25519E2eeClientKeyPair:
    public_key: str
    x25519_secret_key: bytes
    signing_algo: Literal['ed25519'] = 'ed25519'


@dataclass(frozen=True, kw_only=True)
class EcdsaE2eeClientKeyPair:
    public_key: str
    private_key: ec.EllipticCurvePrivateKey
    signing_algo: Literal['ecdsa'] = 'ecdsa'


type E2eeClientKeyPair = Ed25519E2eeClientKeyPair | EcdsaE2eeClientKeyPair


def create_e2ee_client_key_pair(
    signing_algo: SigningAlgo = 'ed25519',
) -> E2eeClientKeyPair:
    """Create a fresh response key; never reuse it between Chat requests."""

    if signing_algo == 'ecdsa':
        private_key = ec.generate_private_key(ec.SECP256K1())
        return EcdsaE2eeClientKeyPair(
            public_key=_ecdsa_public_bytes(private_key)[1:].hex(),
            private_key=private_key,
        )
    key = SigningKey.generate()
    return Ed25519E2eeClientKeyPair(
        public_key=bytes(key.verify_key).hex(),
        x25519_secret_key=bytes(key.to_curve25519_private_key()),
    )


def encrypt_e2ee_text(plaintext: str, model_key: E2eeModelKey) -> str:
    """Encrypt a UTF-8 field for a quote-bound model public key."""

    public_key = hex_to_bytes(model_key.public_key, 'modelSigningPublicKey')
    if model_key.signing_algo == 'ecdsa':
        if len(public_key) == 64:
            public_key = b'\x04' + public_key
        if len(public_key) != 65 or public_key[0] != 4:
            raise verification_failure('e2ee.model_public_key_invalid')
        try:
            recipient = ec.EllipticCurvePublicKey.from_encoded_point(
                ec.SECP256K1(), public_key
            )
        except ValueError as cause:
            raise verification_failure('e2ee.model_public_key_invalid', cause=cause)
        ephemeral = ec.generate_private_key(ec.SECP256K1())
        # ECDH returns the shared point's X coordinate, as in the server protocol.
        key = _derive_key(ephemeral.exchange(ec.ECDH(), recipient), b'ecdsa_encryption')
        nonce = secrets.token_bytes(12)
        ciphertext = AESGCM(key).encrypt(nonce, plaintext.encode(), None)
        return (_ecdsa_public_bytes(ephemeral) + nonce + ciphertext).hex()

    try:
        recipient = bindings.crypto_sign_ed25519_pk_to_curve25519(public_key)
    except (ValueError, RuntimeError) as cause:
        raise verification_failure('e2ee.model_public_key_invalid', cause=cause)
    ephemeral_secret = secrets.token_bytes(32)
    ephemeral_public = bindings.crypto_scalarmult_base(ephemeral_secret)
    key = _derive_key(
        bindings.crypto_scalarmult(ephemeral_secret, recipient), b'ed25519_encryption'
    )
    nonce = secrets.token_bytes(24)
    ciphertext = bindings.crypto_aead_xchacha20poly1305_ietf_encrypt(
        plaintext.encode(), None, nonce, key
    )
    return (ephemeral_public + nonce + ciphertext).hex()


def decrypt_e2ee_text(
    ciphertext: str, client_key_pair: E2eeClientKeyPair, field: str
) -> str:
    """Decrypt a server field, reporting which field failed authentication."""

    # Empty encrypted fields are represented by an empty string on the wire.
    if ciphertext == '':
        return ''
    try:
        envelope = hex_to_bytes(ciphertext, field)
        if isinstance(client_key_pair, EcdsaE2eeClientKeyPair):
            if len(envelope) < 65 + 12 + 16 or envelope[0] != 4:
                raise ValueError('Invalid ECDSA envelope')
            ephemeral = ec.EllipticCurvePublicKey.from_encoded_point(
                ec.SECP256K1(), envelope[:65]
            )
            key = _derive_key(
                client_key_pair.private_key.exchange(ec.ECDH(), ephemeral),
                b'ecdsa_encryption',
            )
            plaintext = AESGCM(key).decrypt(envelope[65:77], envelope[77:], None)
        else:
            if len(envelope) < 32 + 24 + 16:
                raise ValueError('Invalid Ed25519 envelope')
            key = _derive_key(
                bindings.crypto_scalarmult(
                    client_key_pair.x25519_secret_key, envelope[:32]
                ),
                b'ed25519_encryption',
            )
            plaintext = bindings.crypto_aead_xchacha20poly1305_ietf_decrypt(
                envelope[56:], None, envelope[32:56], key
            )
        return plaintext.decode('utf-8')
    except Exception as cause:
        raise verification_failure(
            'e2ee.decryption_failed', {'field': field}, cause=cause
        ) from cause


def _derive_key(shared_secret: bytes, info: bytes) -> bytes:
    return HKDF(algorithm=hashes.SHA256(), length=32, salt=None, info=info).derive(
        shared_secret
    )


def _ecdsa_public_bytes(private_key: ec.EllipticCurvePrivateKey) -> bytes:
    return private_key.public_key().public_bytes(
        serialization.Encoding.X962, serialization.PublicFormat.UncompressedPoint
    )
